### PR TITLE
Final derive output improvements

### PR DESCRIPTION
**Pull request by nnethercote**
With all these changes, the derive output in `deriving-all-codegen.stdout` is pretty close to optimal, i.e. very similar to what you'd write by hand.

r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
Ready for review.

r? @Mark-Simulacrum

**Comment by nnethercote** [comment]
I'm not expecting much change here:

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
@scottmcm: I ended up finding a way to use your `&{self.x}` trick here :)

**Comment by nnethercote** [comment]
The post-merge perf results weren't as good as the pre-merge perf results, but there are still clearly more improvements than regressions, so overall I think it's still a win.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 1]
```diff
@@ -1057,29 +1070,31 @@ impl<'a> MethodDef<'a> {
...
             let selflike_fields =
-                trait_.create_struct_field_access_fields(cx, selflike_args, struct_def);
+                trait_.create_struct_field_access_fields(cx, selflike_args, struct_def, false);
+            mk_body(cx, selflike_fields)
+        } else if always_copy {
+            let selflike_fields =
+                trait_.create_struct_field_access_fields(cx, selflike_args, struct_def, true);
             mk_body(cx, selflike_fields)
         } else {
+            // Neither packed nor copy. Need to use ref patterns.
```
Comment (RalfJung) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1080]: This seems to be a typo? We only get here if `is_packed && !always_copy`.

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 2]
```diff
@@ -1057,29 +1070,31 @@ impl<'a> MethodDef<'a> {
...
             let struct_path = cx.path(span, vec![Ident::new(kw::SelfUpper, type_ident.span)]);
-            let patterns = trait_.create_struct_patterns(
-                cx,
-                struct_path,
-                struct_def,
-                &prefixes,
-                ast::Mutability::Not,
-                use_temporaries,
-            );
+            let use_ref_pat = is_packed && !always_copy;
```
Comment (RalfJung) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1089]: This is always `true`.

Or am I missing something?

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1080]: Good catch. Fixed in #104391.

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 3]
```diff
@@ -1057,29 +1070,31 @@ impl<'a> MethodDef<'a> {
...
+            mk_body(cx, selflike_fields)
+        } else if always_copy {
+            let selflike_fields =
+                trait_.create_struct_field_access_fields(cx, selflike_args, struct_def, true);
             mk_body(cx, selflike_fields)
         } else {
+            // Neither packed nor copy. Need to use ref patterns.
             let prefixes: Vec<_> =
                 (0..selflike_args.len()).map(|i| format!("__self_{}", i)).collect();
+            let addr_of = always_copy;
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1083]: ditto

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1089]: ditto

**Pull request merged**